courseworks/week2/Xuanzhao_Gao/3x+1.ipynb

```diff
@@ -0,0 +1,174 @@
+{
+ "cells": [
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "three_x_solver (generic function with 1 method)"
+      ]
+     },
+     "execution_count": 1,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "function three_x_solver(steps::Int, initial::Int)\n",
+    "    x = Array{Int}(undef, steps)\n",
+    "    p_step = initial\n",
+    "    for s in 1:steps\n",
+    "        if isodd(p_step) == 1\n",
+    "            p_step *= 3\n",
+    "            p_step += 1\n",
+    "        else\n",
+    "            p_step /= 2\n",
+    "        end\n",
+    "        x[s] = p_step\n",
+    "    end\n",
+    "    return x\n",
+    "end"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "three_x_add_one_solver (generic function with 1 method)"
+      ]
+     },
+     "execution_count": 2,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "function three_x_add_one_solver(initial::Int)\n",
+    "    p_step = initial\n",
+    "    step = 0\n",
+    "    while p_step != 1\n",
```

Comment:
To make the program faster, you can use
```julia
while p_step >= initial
```

As the break condition. Then you may get all `Int32` numbers tested in 1min.
```julia
julia> @time @testset "test through INT32" begin
           res = true
           for i in 2:Int(typemax(Int32))
               res &= three_x_add_one_solver(i) == 1
           end
           @test res
       end
Test Summary:      | Pass  Total   Time
test through INT32 |    1      1  58.2s
 58.199649 seconds (246 allocations: 11.070 KiB)
Test.DefaultTestSet("test through INT32", Any[], 1, false, false, true, 1.67904979576294e9, 1.679049853962387e9)
```